src/serializer/ResidualHeapVisitor.js

```diff
@@ -1211,8 +1211,7 @@ export class ResidualHeapVisitor {
         // This may not have been referentialized if the binding is a local of an optimized function.
         // in that case, we need to figure out which optimized function it is, and referentialize it in that scope.
         let optimizedFunctionScope = this._getAdditionalFunctionOfScope();
-        if (residualBinding.potentialReferentializationScopes.size === 0) {
-          invariant(optimizedFunctionScope !== undefined);
+        if (residualBinding.potentialReferentializationScopes.size === 0 && optimizedFunctionScope !== undefined) {
           this._enqueueWithUnrelatedScope(optimizedFunctionScope, () => {
             invariant(additionalFunctionInfo !== undefined);
```

Comment:
Won't this only be true if we're inside an optimized function? Also not sure if `_enqueueWithUnrelatedScope` does something reasonable when passed `undefined`.

Reply:
So looking back at why I added those bits of code, they're only an issue for optimized functions (it prevents a case where something is only reachable from an optimized function and we didn't end up visiting it). We should be fine if it has to do with global/residual code.

I guess what I was trying to say is that I'm surprised that we use visitBindingAssignment outside of optimized function cases (I wasn't aware of `evaluatePureFunction`) and that it wasn't designed for that, and that I wasn't sure that it would do the right thing.